flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/CastRuleUtils.java

```diff
@@ -163,6 +166,25 @@ static String binaryWriterWriteNull(
         return CodeGenUtils.binaryWriterWriteNull(indexTerm, writerTerm, logicalType);
     }
 
+    static String lengthExceedsPrecision(String strTerm, int precision) {
+        return methodCall(strTerm, "length") + " > " + precision;
+    }
+
+    /**
+     * For complex types such as {@link LogicalTypeRoot#ARRAY}, use {@link VarCharType#STRING_TYPE}
+     * to avoid trimming/padding when casting the inner elements to a {@link
+     * LogicalTypeRoot#VARCHAR} or {@link LogicalTypeRoot#CHAR}.
+     */
+    static LogicalType getLogicalTypeForInnerCast(LogicalType logicalType) {
+        return logicalType.is(LogicalTypeFamily.CHARACTER_STRING)
+                ? VarCharType.STRING_TYPE
+                : logicalType;
+    }
```

Comment:
Why you need this? In every call site you use it, you're always casting to `CHARACTER_STRING`, so you'll never fall back to the original type

Reply:
It was more of "safety net" but you're right we don't need it at all, I will directly call underlying code gen with `STRING_TYPE`